// aws-sdk-client-mock/dist/es/awsClientStub.js
import { match } from 'sinon';
/**
 * Wrapper on the mocked `Client#send()` method,
 * allowing to configure its behavior.
 *
 * Without any configuration, `Client#send()` invocation returns `undefined`.
 *
 * To define resulting variable type easily, use {@link AwsClientStub}.
 */
var AwsStub = /** @class */ (function () {
    function AwsStub(send) {
        this.send = send;
        this.anyCommandBehavior = new CommandBehavior(this, send);
    }
    /** Resets stub's history and behavior. */
    AwsStub.prototype.reset = function () {
        this.send.reset();
        return this;
    };
    /** Resets stub's behavior. */
    AwsStub.prototype.resetBehavior = function () {
        this.send.resetBehavior();
        return this;
    };
    /** Resets stub's calls history. */
    AwsStub.prototype.resetHistory = function () {
        this.send.resetHistory();
        return this;
    };
    /** Replaces stub with original `Client#send()` method. */
    AwsStub.prototype.restore = function () {
        this.send.restore();
    };
    /**
     * Returns recorded calls to the stub.
     * Clear history with {@link resetHistory} or {@link reset}.
     */
    AwsStub.prototype.calls = function () {
        return this.send.getCalls();
    };
    /**
     * Returns n-th recorded call to the stub.
     */
    AwsStub.prototype.call = function (n) {
        return this.send.getCall(n);
    };
    /**
     * Returns recorded calls of given Command only.
     * @param commandType Command type to match
     * @param input Command payload to match
     * @param strict Should the payload match strictly (default false, will match if all defined payload properties match)
     */
    AwsStub.prototype.commandCalls = function (commandType, input, strict) {
        var _this = this;
        return this.send.getCalls()
            .filter(function (call) {
            var isProperType = call.args[0] instanceof commandType;
            var inputMatches = _this.createInputMatcher(input, strict).test(call.args[0]);
            return isProperType && inputMatches;
        });
    };
    /**
     * Allows specifying the behavior for a given Command type and its input (parameters).
     *
     * If the input is not specified, it will match any Command of that type.
     * @param command Command type to match
     * @param input Command payload to match
     * @param strict Should the payload match strictly (default false, will match if all defined payload properties match)
     */
    AwsStub.prototype.on = function (command, input, strict) {
        if (strict === void 0) { strict = false; }
        var matcher = match.instanceOf(command).and(this.createInputMatcher(input, strict));
        var cmdStub = this.send.withArgs(matcher);
        return new CommandBehavior(this, cmdStub);
    };
    /**
     * Allows specifying the behavior for any Command with given input (parameters).
     *
     * If the input is not specified, the given behavior will be used for any Command with any input.
     * This is no different from using {@link resolves}, {@link rejects}, etc. directly,
     * but can be used for readability.
     * @param input Command payload to match
     * @param strict Should the payload match strictly (default false, will match if all defined payload properties match)
     */
    AwsStub.prototype.onAnyCommand = function (input, strict) {
        if (strict === void 0) { strict = false; }
        var cmdStub = this.send.withArgs(this.createInputMatcher(input, strict));
        return new CommandBehavior(this, cmdStub);
    };
    AwsStub.prototype.createInputMatcher = function (input, strict) {
        if (strict === void 0) { strict = false; }
        return input !== undefined ?
            match.has('input', strict ? input : match(input))
            : match.any;
    };
    /**
     * Sets a successful response that will be returned from any `Client#send()` invocation.
     * @param response Content to be returned
     */
    AwsStub.prototype.resolves = function (response) {
        return this.anyCommandBehavior.resolves(response);
    };
    /**
     * Sets a failure response that will be returned from any `Client#send()` invocation.
     * The response will always be an `Error` instance.
     * @param error Error text, Error instance or Error parameters to be returned
     */
    AwsStub.prototype.rejects = function (error) {
        return this.anyCommandBehavior.rejects(error);
    };
    /**
     * Sets a function that will be called on any `Client#send()` invocation.
     * @param fn Function taking Command input and returning result
     */
    AwsStub.prototype.callsFake = function (fn) {
        return this.anyCommandBehavior.callsFake(fn);
    };
    return AwsStub;
}());
export { AwsStub };
var CommandBehavior = /** @class */ (function () {
    function CommandBehavior(clientStub, send) {
        this.clientStub = clientStub;
        this.send = send;
    }
    /**
     * Sets a successful response that will be returned from the `Client#send()` invocation
     * for specified Command and/or its input.
     * @param response Content to be returned
     */
    CommandBehavior.prototype.resolves = function (response) {
        this.send.resolves(response);
        return this.clientStub;
    };
    /**
     * Sets a failure response that will be returned from the `Client#send()` invocation
     * for specified Command and/or its input.
     * The response will always be an `Error` instance.
     * @param error Error text, Error instance or Error parameters to be returned
     */
    CommandBehavior.prototype.rejects = function (error) {
        if (typeof error === 'string') {
            error = new Error(error);
        }
        if (!(error instanceof Error)) {
            error = Object.assign(new Error(), error);
        }
        this.send.rejects(error);
        return this.clientStub;
    };
    /**
     * Sets a function that will be called on `Client#send()` invocation
     * for specified Command and/or its input.
     * @param fn Function taking Command input and returning result
     */
    CommandBehavior.prototype.callsFake = function (fn) {
        this.send.callsFake(function (cmd) { return fn(cmd.input); });
        return this.clientStub;
    };
    return CommandBehavior;
}());
export { CommandBehavior };
